### DISCUSSION TITLE
questions on k3s hardening

### DISCUSSION THREAD

**Discussion by cwrx777** [k3s-io/k3s#10458] [Q&A]
1. where does k3s set the default [control plane execution and arguments](https://docs.k3s.io/security/hardening-guide#control-plane-execution-and-arguments) for apiserver, controller manager, scheduler, kubelet?
2. Any reasons why the required Admission control plugins are not by default? is it supposed to be set in my config.yaml, like [this](https://docs.k3s.io/security/hardening-guide#configuration-for-kubernetes-components) ?
  - [NamespaceLifecycle](https://docs.k3s.io/security/hardening-guide#control-1215), 
  - [PodSecurityPolicy](https://docs.k3s.io/security/hardening-guide#control-1216)
  - [EventRateLimit](https://docs.k3s.io/security/self-assessment-1.8#129-ensure-that-the-admission-control-plugin-eventratelimit-is-set-manual)
  - [AlwaysPullImages](https://docs.k3s.io/security/self-assessment-1.8#129-ensure-that-the-admission-control-plugin-eventratelimit-is-set-manual)
3. In case I need to enable additional admission plugins or other k8s component parameters using my config.yaml, do I need to also specify the default value, e.g. `NodeRestriction`? 
4. What does [this ](https://docs.k3s.io/security/self-assessment-1.8#1214-ensure-that-the-admission-control-plugin-namespacelifecycle-is-set-automated)command do?
   ```bash
   journalctl -D /var/log/journal  -u k3s | grep 'Running kube-apiserver' | tail -n1 | grep -v grep
   ```
5. Is [configuration for kubernetes components](https://docs.k3s.io/security/hardening-guide#configuration-for-kubernetes-components) content outdated?
6. is [RotateKubeletServerCertificate ](https://docs.k3s.io/security/self-assessment-1.8#136-ensure-that-the-rotatekubeletservercertificate-argument-is-set-to-true-automated)supported in v1.29?

**Comment by brandond** [comment]
Please don't open multiple discussions on the same topic. Lets keep this at https://github.com/k3s-io/docs/issues/300#issuecomment-2214864206

**Reply by cwrx777** [reply]
@brandond
possible to use this thread instead ?

**Comment by dereknola** [comment]
1. Most are set in https://github.com/k3s-io/k3s/blob/master/pkg/daemons/control/server.go
2. K3s is not hardened by default. You should look into [RKE2](https://docs.rke2.io/), which has hardening with `profile:cis` if you are interested in that. Also, PSP are only used on 1.24 and older. That Known Issues section needs to be cleaned up.
3. I'm not sure what you mean by this. Any additional parameters you set are on you to support.
4. K3s runs all the kubernetes control plane components in a single process. This is looking thru the process logs to find the arguments that the kube-apiserver launched with. Try and run it yourself on your nodes and you will see what I'm talking about.
5. No, you should be able to pass up through cis-1.8 with that configuration. cis-1.9 just hit kube-bench and we are working on getting K3s/RKE2 conformant with it (although not much will likely change, we just haven't 100% finished that work).
6. Yes, this check is outdated, as the rotate feature gate has been "on by default" since Kubernetes 1.11. This check is basically just "did you do something stupid and turn off the security feature".

**Reply by cwrx777** [reply]
Hi @dereknola Thank you for answering my questions. I have additoinal questions.

> I'm not sure what you mean by this. Any additional parameters you set are on you to support.

After I run k3s server with `config.yaml` according to [this](https://docs.k3s.io/security/hardening-guide#configuration-for-kubernetes-components), I can see that there are parameters that are set even it is not in the configuration file. and you have answered that most are set in [here](https://github.com/k3s-io/k3s/blob/master/pkg/daemons/control/server.go).  
Using [`enable-admission-plugins`](https://github.com/k3s-io/k3s/blob/master/pkg/daemons/control/server.go#L229) as example, it is hardcoded to `NodeRestriction`. If I want to enable additional plugins, I believe i should put that in the k3s config.yaml.
Do I put it as 
`enable-admission-plugins=NodeRestriction,EventRateLimit` or
`enable-admission-plugins=EventRateLimit` ?


> Yes, this check is outdated, as the rotate feature gate has been "on by default" since Kubernetes 1.11. This check is basically just "did you do something stupid and turn off the security feature".

how do I audit that `RotateKubeletServerCertificate` feature is enabled? 

Additional questions:

- [1.1.13](https://docs.k3s.io/security/self-assessment-1.8#1113-ensure-that-the-adminconf-file-permissions-are-set-to-600-or-more-restrictive-automated)
  why is this `Not Applicable` ? isn't it just changing the file permission to 600?

My worker node is run using the following `config.yaml`:
```yaml
...
protect-kernel-defaults: true
kubelet-arg:
  - 'streaming-connection-idle-timeout=5m'
  - 'make-iptables-util-chains=true'
```
`k3s-agent.service`:
```
ExecStart=/usr/local/bin/k3s \
    agent \
        '--config=/test/k3s/config/config.yaml' \
```

- [4.1.9](https://docs.k3s.io/security/self-assessment-1.8#419-ensure-that-the-kubelet---config-configuration-file-has-permissions-set-to-600-or-more-restrictive-automated), [4.1.10](https://docs.k3s.io/security/self-assessment-1.8#4110-ensure-that-the-kubelet---config-configuration-file-ownership-is-set-to-root-automated)
  Does [`kubelet --config configuration file`](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#:~:text=%2D%2Dconfig%20string) refer to [`kubelet.kubeconfig`](https://kubernetes.io/docs/reference/command-line-tools-reference/kubelet/#:~:text=Path%20to%20a-,kubeconfig,-file%2C%20specifying%20how)?
- [4.2.8](https://docs.k3s.io/security/self-assessment-1.8#428-ensure-that-the-eventrecordqps-argument-is-set-to-a-level-which-ensures-appropriate-event-capture-manual), [4.2.10](https://docs.k3s.io/security/self-assessment-1.8#4210-ensure-that-the---rotate-certificates-argument-is-not-set-to-false-manual), [4.2.11](https://docs.k3s.io/security/self-assessment-1.8#4211-verify-that-the-rotatekubeletservercertificate-argument-is-set-to-true-manual) 
   I am not able to verify the required parameters from the ps output. 
   ```bash
  /bin/ps -fC containerd
   UID          PID    PPID  C STIME TTY          TIME CMD
   root     3002164 3001447  0 Jul08 ?        00:06:39 containerd
   ```
- [4.2.12](https://docs.k3s.io/security/self-assessment-1.8#4212-ensure-that-the-kubelet-only-makes-use-of-strong-cryptographic-ciphers-manual) 
 Does the kubelet config refer to `/var/lib/rancher/k3s/agent/kubelet.kubeconfig `
`kubelet.kubeconfig`:
  ```
  apiVersion: v1
  clusters:
  - cluster:
      server: https://127.0.0.1:6444
      certificate-authority: /var/lib/rancher/k3s/agent/server-ca.crt
    name: local
  contexts:
  - context:
      cluster: local
      namespace: default
      user: user
    name: Default
  current-context: Default
  kind: Config
  preferences: {}
  users:
  - name: user
    user:
      client-certificate: /var/lib/rancher/k3s/agent/client-kubelet.crt
      client-key: /var/lib/rancher/k3s/agent/client-kubelet.key
  ```
- [4.2.1.3](https://docs.k3s.io/security/self-assessment-1.8#4213-ensure-that-a-limit-is-set-on-pod-pids-manual)
  Where is `pod-pids` supposed to be set? I dont see that this is set in the kubelet CLI arguments. 
   ```bash
    Running kubelet 
  --address=0.0.0.0 
  --allowed-unsafe-sysctls=net.ipv4.ip_forward,net.ipv6.conf.all.forwarding 
  --anonymous-auth=false 
  --authentication-token-webhook=true 
  --authorization-mode=Webhook 
  --cgroup-driver=systemd 
  --client-ca-file=/var/lib/rancher/k3s/agent/client-ca.crt 
  --cloud-provider=external 
  --cluster-dns=10.43.0.10 
  --cluster-domain=cluster.local 
  --container-runtime-endpoint=unix:///run/k3s/containerd/containerd.sock 
  --containerd=/run/k3s/containerd/containerd.sock 
  --eviction-hard=imagefs.available<5%,nodefs.available<5% 
  --eviction-minimum-reclaim=imagefs.available=10%,nodefs.available=10% 
  --fail-swap-on=false 
  --feature-gates=CloudDualStackNodeIPs=true 
  --healthz-bind-address=127.0.0.1 
  --hostname-override=skhromivpapp03.shses.shs.com.sg 
  --kubeconfig=/var/lib/rancher/k3s/agent/kubelet.kubeconfig 
  --make-iptables-util-chains=true 
  --node-ip=10.236.49.161 
  --node-labels=node-type=worker 
  --pod-infra-container-image=rancher/mirrored-pause:3.6 
  --pod-manifest-path=/var/lib/rancher/k3s/agent/pod-manifests 
  --protect-kernel-defaults=true 
  --read-only-port=0 
  --resolv-conf=/etc/resolv.conf 
  --serialize-image-pulls=false 
  --streaming-connection-idle-timeout=5m 
  --tls-cert-file=/var/lib/rancher/k3s/agent/serving-kubelet.crt 
  --tls-private-key-file=/var/lib/rancher/k3s/agent/serving-kubelet.key
     ```
